src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.InFlightSolution.cs

```diff
@@ -100,29 +102,43 @@ public void IncrementInFlightCount_NoLock()
                 InFlightCount++;
             }
 
-            public void DecrementInFlightCount_NoLock()
+            /// <summary>
+            /// Returns the in-flight solution computations <em>when</em> the in-flight-count is decremented to 0. This
+            /// allows the caller to wait for those computations to complete (which will hopefully be quickly as they
+            /// will have just been canceled).  This ensures the caller doesn't return back to the host (potentially
+            /// unpinning the solution on the host) while the solution-computation tasks are still running and may still
+            /// attempt to call into the host.
+            /// </summary>
+            public ImmutableArray<Task<Solution>> DecrementInFlightCount_NoLock()
```

Comment:
should this instead return a non-generic Task that combines the two solutions (does Task.WhenAll work here?)  That way a caller can't access the solution objects themselves.

Reply:
talked offline.  reason for this written above.